lib/hotel.rb

```diff
@@ -0,0 +1,184 @@
+						date.start_date < current_block.end_date
+							raise ArgumentException, "This room is part of a block. Change your start date"
+						elsif date.end_date > current_block.start_date
+							raise ArgumentException, "This room is part of a block. Change your end date"
+						end
+					end
+				end
+			end
+		end
+		
+		#pulls all array of all reservations of a given room
+		current_reservations = @reservations_by_room[room_number]
+		current_reservations.each do |i|
+			#if the requested start date is before the current reservations start date
+			if date.start_date < i.start_date  
+				#checks to see if the end date is also before the start date
+				if (date.end_date - i.start_date).to_i > 0 
+					raise ArgumentError, "This room is booked during the dates requested"
+				end
+				#if the start dates are the same, or the start date is less than the end date
+			elsif date.start_date == i.start_date || (date.start_date - i.end_date).to_i < 0 
+				raise ArgumentError, "This room is booked during the dates requested"
+			end
+		end
+		
+		#makes new reservation
+		new_reservation = Hotels::Reservation.new(room_number, date)
+		@reservations_by_room[room_number] << new_reservation
+		
+		#if this day already has a reservation
+		if @reservations_by_date.key?(date.start_date)
+			#add it to the array
+			@reservations_by_date[date.start_date] << new_reservation
+		else
+			#other wise make a new array and add it
+			@reservations_by_date[date.start_date] = [new_reservation]
+		end
+			
+	  end
+	  
+	  
+	  def get_all_rooms
+		return @rooms
+	  end
+	  
+	  def get_open_rooms(date)
+		open_rooms = []
+		#makes a boolean array flagging all rooms as available
+		20.times do |i|
+			open_rooms[i] = true
+		end
+		
+		current_date = date.start_date
+		while current_date < date.end_date do
+			#if the current_date has a reservation
+			if @reservations_by_date.key?(current_date)
+				current_reservations = @reservations_by_date[current_date]
+				#iterate over each reservation
+				current_reservations.each do  |this_reservation|
+					#set value of the boolean array (where the index == room number) to false
+					open_rooms[this_reservation.room_number] = false
+				end
+			end
+			current_date += 1
+		end
+		
+		#this will hold the actual avaialble rooms
+		available_rooms = []
+		index = 0
+		
+		open_rooms.each do |value|
+			#if the value of the boolean array is true
+			if value == true
+				#append the room at the index to the available_rooms array
+				available_rooms << @rooms[index]
+			end
+			index += 1
+		end
+		
+		return available_rooms
+	  end
+	  
+	  def make_new_block(number_of_rooms, date, id)
+		available_rooms = get_open_rooms(date)
+		if number_of_rooms > 5
+			raise ArgumentException, "Room blocks must be 5 rooms or less"
+		elsif number_of_rooms > available_rooms.length
+			raise ArgumentException, "There are only #{available_rooms.length} available for the dates selected"
```

Comment:
Because both of these first two cases end in a `raise` statement we can assume that if either of them are hit we won't continue running the rest of this method. This means we don't need to contain the actual "meat" of the method's logic inside of an `else` block.